src/map.rs

```diff
@@ -492,30 +494,37 @@ where
         self.map.clear()
     }
 
-    /// Update entry's expiration status and return old status.
-    pub fn update_expiration_status(&mut self, key: K, duration: Duration) -> Option<EntryStatus> {
-        let mut old_status: Option<EntryStatus> = None;
-        {
-            if let Some(entry) = self.map.get_mut(&key) {
-                let old_status_ = Some(*entry.status());
-                let expiration_secs = duration.as_secs();
-                entry.update_status(EntryStatus::ExpiresAtSeconds(expiration_secs));
+    /// Updates the expiration status of an entry and returns the old status.
+    ///
+    /// If the entry does not exist, returns `Err(TimedMapError::EntryNotFound)`.
+    /// If the entry's old status is `EntryStatus::Constant`, returns None
+    pub fn update_expiration_status(
+        &mut self,
+        key: K,
+        duration: Duration,
+    ) -> Result<Option<EntryStatus>, TimedMapError> {
+        match self.map.get_mut(&key) {
+            Some(entry) => {
+                let old_status = *entry.status();
+                let now = self.clock.elapsed_seconds_since_creation();
+                let expires_at = now + duration.as_secs();
+                entry.update_status(EntryStatus::ExpiresAtSeconds(expires_at));
 
                 self.expiries
-                    .entry(expiration_secs)
+                    .entry(expires_at)
                     .or_default()
                     .insert(key.clone());
 
-                old_status = old_status_;
+                match &old_status {
+                    EntryStatus::Constant => Ok(None),
+                    EntryStatus::ExpiresAtSeconds(t) => {
+                        self.drop_key_from_expiry(t, &key);
+                        Ok(Some(old_status))
+                    }
+                }
             }
+            None => Err(TimedMapError::EntryNotFound),
```

Comment:
Let's use `NotFound` from std io to avoid creating custom error type for one function.

Reply:
we need to serve `no_std` too?

Reply:
It is, right... I don't know what to do here (creating an error type for one function seem trivial to me). Do whatever you feel is best.

Reply:
removed custom error and instead returning `static str`